<launch>
  <!-- HoloOcean Configuration -->
  <arg name="env_config" default="OceanSimple-Hovering2CameraOnly-VisibilityEstimation"/>
  <arg name="show_viewport" default="true"/>
  <arg name="fog_density" default="2.0"/>
  <arg name="render_quality" default="3"/>
  <arg name="trajectory_update_rate" default="10.0"/>
  <!-- Path to config files -->
  <arg name="pid_config" default="$(find holoocean_ros)/config/pid_config.yaml"/>
  <arg name="camera_intrinsics_file" default="$(find holoocean_ros)/config/camera_intrinsics.yaml"/>
  <!-- Launch the HoloOcean ROS node -->
  <node name="holoocean_integrated" pkg="holoocean_ros" type="holoocean_node.py" output="screen">
    <param name="env_config" value="$(arg env_config)"/>
    <param name="show_viewport" value="$(arg show_viewport)"/>
    <param name="fog_density" value="$(arg fog_density)"/>
    <param name="render_quality" value="$(arg render_quality)"/>
    <param name="trajectory_update_rate" value="$(arg trajectory_update_rate)"/>
    <param name="pid_config" value="$(arg pid_config)"/>
    <param name="camera_intrinsics_file" value="$(arg camera_intrinsics_file)"/>
    <!--RAFT stereo remap for odometry -->
    <remap from="/odometry" to="/rovio/odometry"/>
  </node>
  <!-- Pose converter for HoloOcean - Modified to handle PoseWithCovarianceStamped -->
  <node pkg="holoocean_ros" type="pose_converter.py" name="pose_converter" output="screen">
    <!-- Make sure the pose_converter publishes PoseWithCovarianceStamped -->
    <remap from="/holoocean/current_pose" to="/holoocean/raw_pose"/>
  </node>
  
  <!-- Start nodelet manager -->
  <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" output="screen"/>
  <!-- Start depth_image_proc nodelet -->
  <node pkg="nodelet" type="nodelet" name="depth_to_pointcloud"
        args="load depth_image_proc/point_cloud_xyz nodelet_manager" output="screen">
    <remap from="image_rect" to="/camera/depth/image_raw"/>
    <remap from="camera_info" to="/camera/depth/camera_info"/>
    <remap from="points" to="/camera/depth/points"/>
    <param name="sensor_frame" value="camera_depth_optical_frame" />
  </node>
  
  <!-- Odometry republisher to ensure we have odometry messages -->
  <node pkg="topic_tools" type="relay" name="odometry_relay" 
        args="/rovio/odometry /odometry" />
</launch>